gui/src/app/pyodide/pyodideWorker/pyodideWorker.ts

```diff
@@ -0,0 +1,246 @@
+    }
+    if (pyodideWorkerMode === "analysis.py") {
+      const sp_load_draws_py = await fetchScript('/sp_load_draws.py')
+      pyodide.FS.writeFile("sp_load_draws.py", sp_load_draws_py, {
+        encoding: "utf-8",
+      });
+    }
+    return pyodide;
+  } else {
+    return pyodide;
+  }
+};
+
+self.onmessage = (e) => {
+  const message = e.data;
+  if (!isMessageToPyodideWorker(message)) {
+    console.error("Invalid message from main", message);
+    return;
+  }
+  if (message.type === "setPyodideWorkerMode") {
+    if (pyodideWorkerMode !== undefined) {
+      throw Error("pyodideWorkerMode is already defined");
+    }
+    if (!["data.py", "analysis.py"].includes(message.mode)) {
+      throw Error("Invalid pyodideWorkerMode");
+    }
+    pyodideWorkerMode = message.mode;
+  } else if (message.type === "run") {
+    run(message.code, message.spData);
+  }
+};
+
+const sendMessageToMain = (message: MessageFromPyodideWorker) => {
+  self.postMessage(message);
+};
+
+const sendStdout = (data: string) => {
+  sendMessageToMain({ type: "stdout", data });
+};
+
+const sendStderr = (data: string) => {
+  sendMessageToMain({ type: "stderr", data });
+};
+
+const setStatus = (status: PydodideWorkerStatus) => {
+  sendMessageToMain({ type: "setStatus", status });
+};
+
+const setData = (data: any) => {
+  if (pyodideWorkerMode !== "data.py") {
+    throw Error("setData is only supported in data.py mode");
+  }
+  sendMessageToMain({ type: "setData", data });
+};
+
+const addImage = (image: any) => {
+  if (pyodideWorkerMode !== "analysis.py") {
+    throw Error("addImage is only supported in analysis.py mode");
+  }
+  sendMessageToMain({ type: "addImage", image });
+};
+
+const run = async (code: string, spData: any) => {
+  if (!pyodideWorkerMode) {
+    throw Error("pyodideWorkerMode is not defined");
+  }
+  setStatus("loading");
+  try {
+    const pyodide = await loadPyodideInstance();
+    setStatus("running");
+
+    const scriptPreamble = getScriptPreable(pyodideWorkerMode);
+
+    // here's where we can pass in globals
+    const globals = pyodide.toPy({ _stan_playground: true });
+    let script = scriptPreamble + "\n" + code;
+
+    if (pyodideWorkerMode === "data.py") {
+      // We serialize the data object to json string in the python script
+      script += "\n";
+      script += "import stanio\n";
+      script += "import json\n";
+      script += "data = stanio.dump_stan_json(data)\n";
+    }
+
+    let succeeded = false;
+    try {
+      if (script.includes("arviz")) {
+        // If the script has arviz, we need to install it
+        setStatus("loading");
+        try {
+          await pyodide.loadPackage("micropip");
+          const microPip = pyodide.pyimport("micropip");
+          await microPip.install("arviz<0.18");
+        } finally {
+          setStatus("running");
+        }
+      }
+      pyodide.FS.writeFile("_sp_data.json", JSON.stringify(spData), {
+        encoding: "utf-8",
+      });
```

Comment:
Writing this to a file seems a bit odd to me compared to initializing it as a global. Is there an advantage to this approach?

Reply:
I think it's cleaner as a file because then any of the scripts can access it without needing to pass it around as a variable. And I think it's good to keep the main script as simple as possible and put more complex logic in the sp_load_draws.py

Reply:
@WardBrian I thought about it some more and I think you're right. Now it's being passed as global variable. (but I still think sp_load_draws.py should be in its own file)